pyupgrade.py

```diff
@@ -1537,6 +1535,37 @@ def generic_visit(self, node):  # type: (ast.AST) -> None
         super(FindPy3Plus, self).generic_visit(node)
 
 
+def reversed_walk(node):  # type: (ast.AST) -> Generator[ast.AST, None, None]
+    todo = [node]
+    while todo:
+        node = todo.pop()
+        todo.extend(ast.iter_child_nodes(node))
+        yield node
+
+
+def targets_referenced_after_node(root_node, node, targets):
+    # type: (ast.AST, ast.AST, Union[ast.expr, ast.Tuple]) -> bool
+    """
+    Search inside the given root node whether the targets have been referenced
+    after the given node. Walks from behind until the node have been reached.
+    """
+    for x in reversed_walk(root_node):
+        # Finish walking when the node has been reached
+        if x == node:
+            break
+
+        # Filter for name expressions
+        if not isinstance(x, ast.Name):
+            continue
+
+        # Match one of the targets
+        for t in ast.walk(targets):
+            if fields_same(t, x):
+                return True
```

Comment:
It would probably be better to refactor to this (above the loop)

```python
names = {node.id for node in ast.walk(targets)}

...

if isinstance(x, ast.Name) and isinstance(x.ctx, ast.Load) and x.id in names:
    ...
```

Reply:
```python
names = {node.id for node in ast.walk(targets)}
```
That's neat, but it seems it doesn't handle targets as `ast.Tuple` (`for x, y in z: yield x, y` case). That's why there's `fields_same`.

Reply:
Maybe, it's better to walk through the targets and compose the set with node.id.

Reply:
it's missing an `if isinstance(node, ast.Name)` -- that's what I get for phone

Reply:
Yay. that's working!

Reply:
Addressed in 81dc036f6a52e32a3dd1e992334c24fc09cea12b.